crates/iceberg-catalog/src/service/task_queue/tabular_expiration_queue.rs

```diff
@@ -0,0 +1,334 @@
+            e
+        })?;
+    tracing::debug!("Got FileIO");
+
+    handle_old_metadata(
+        warehouse_ident,
+        &fio,
+        &table_metadata.table_metadata,
+        delete_queue,
+        task_id,
+    )
+    .await
+    .map_err(|e| {
+        tracing::error!("Failed to handle old metadata: {:?}", e);
+        e
+    })?;
+
+    let location = table_metadata.metadata_location.as_deref().ok_or_else(|| {
+        let err = ErrorModel::internal(
+            "Table metadata location not found",
+            "MetadataLocationNotFound",
+            None,
+        );
+        tracing::error!("Table metadata location not found: {:?}", err);
+        err
+    })?;
+
+    find_files(
+        &table_metadata.table_metadata,
+        location,
+        warehouse_ident,
+        delete_queue,
+        &fio,
+        task_id,
+    )
+    .await
+    .map_err(|e| {
+        tracing::error!("Failed to find files: {:?}", e);
+        e
+    })?;
+
+    C::drop_table(
+        tabular_id.into(),
+        DropFlags {
+            hard_delete: true,
+            purge: true,
+        },
+        trx.transaction(),
+    )
+    .await
+    .map_err(|e| {
+        tracing::error!("Failed to drop table: {:?}", e);
+        e
+    })?;
+
+    trx.commit().await.map_err(|e| {
+        tracing::error!("Failed to commit transaction: {:?}", e);
+        e
+    })?;
+
+    Ok(())
+}
+
+async fn handle_old_metadata(
+    warehouse_id: WarehouseIdent,
+    fio: &FileIO,
+    meta: &TableMetadata,
+    fetcher: &impl TaskQueue<Task = Deletion, Input = DeleteInput>,
+    task_id: Uuid,
+) -> Result<()> {
+    for old_meta in &meta.metadata_log {
+        // old metadata doesnt exist in db hence we're going to s3
+        let input = fio
+            .new_input(old_meta.metadata_file.as_str())
+            .map_err(|e| {
+                ErrorModel::internal("Failed to create input", "FileIOError", Some(Box::new(e)))
+            })?;
+        let metadata = input.read().await.map_err(|e| {
+            ErrorModel::internal("Failed to read input", "FileIOError", Some(Box::new(e)))
+        })?;
+        let meta = serde_json::from_slice::<TableMetadata>(&metadata).map_err(|e| {
+            ErrorModel::internal(
+                "Failed to deserialize metadata",
+                "SerdeJsonError",
+                Some(Box::new(e)),
+            )
+        })?;
+
+        find_files::<_>(
+            &meta,
+            old_meta.metadata_file.as_str(),
+            warehouse_id,
+            fetcher,
+            fio,
+            task_id,
+        )
+        .await?;
+    }
+    Ok(())
+}
+
+async fn find_files<D: TaskQueue<Task = Deletion, Input = DeleteInput>>(
+    metadata: &TableMetadata,
+    metadata_location: &str,
+    warehouse_ident: WarehouseIdent,
+    conn: &D,
+    fio: &FileIO,
+    task_id: Uuid,
+) -> Result<()> {
+    for file in metadata.snapshots.values() {
+        let list = file.load_manifest_list(fio, metadata).await.map_err(|e| {
```

Comment:
If we don't switch to my preferred drop mechanism to just remove the whole directory recursively, we need a caching mechanism here: Tables can have a huge number of snapshots that all use the same manifest list as part of the snapshot. Revisiting it is very costly and adds a massive number of redundant io operations.